photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/CoordinateFactory.scala

```diff
@@ -34,35 +37,40 @@ import com.linkedin.photon.ml.util.PhotonBroadcast
 object CoordinateFactory {
 
   /**
-   * Creates a [[Coordinate]] of the appropriate type, given the input [[Dataset]],
+   * Creates a [[Coordinate]] of the appropriate type, given the input data set,
    * [[CoordinateOptimizationConfiguration]], and [[ObjectiveFunction]].
    *
-   * @tparam D Some type of [[Dataset]]
    * @param dataset The input data to use for training
+   * @param featureShardId
+   * @param inputColumnsNames
    * @param coordinateOptConfig The optimization settings for training
    * @param lossFunctionFactoryConstructor A constructor for the loss function factory function
    * @param glmConstructor A constructor for the type of [[GeneralizedLinearModel]] being trained
    * @param downSamplerFactory A factory function for the [[DownSampler]] (if down-sampling is enabled)
    * @param normalizationContext The [[NormalizationContext]]
    * @param varianceComputationType Should the trained coefficient variances be computed in addition to the means?
    * @param interceptIndexOpt The index of the intercept, if one is present
-   * @return A [[Coordinate]] for the [[Dataset]] of type [[D]]
+   * @param rETypeOpt
+   * @return A [[Coordinate]] instance
    */
-  def build[D <: Dataset[D]](
-      dataset: D,
+  def build(
+      dataset: DataFrame,
+      featureShardId: FeatureShardId,
+      inputColumnsNames: InputColumnsNames,
       coordinateOptConfig: CoordinateOptimizationConfiguration,
       lossFunctionFactoryConstructor: ObjectiveFunctionFactoryFactory,
       glmConstructor: Coefficients => GeneralizedLinearModel,
       downSamplerFactory: DownSamplerFactory,
       normalizationContext: NormalizationContext,
       varianceComputationType: VarianceComputationType,
-      interceptIndexOpt: Option[Int]): Coordinate[D] = {
+      interceptIndexOpt: Option[Int],
+      rETypeOpt: Option[REType]): Coordinate = {
 
     val lossFunctionFactory = lossFunctionFactoryConstructor(coordinateOptConfig)
 
-    (dataset, coordinateOptConfig, lossFunctionFactory) match {
+    (rETypeOpt, coordinateOptConfig, lossFunctionFactory) match {
```

Comment:
Can we just do a match on (`coordinateOptConfig`, `lossFunctionFactory`)? This `rETypeOpt` seems to be redundant.